### PR TITLE
Feature/162416247 make search results titles clickable to sort

### PR DESCRIPTION
**Pull request by lingyun1010**
I pass the header array with css style as a prop into the facet search results repository, hence, the selected sorting header can set states and update table content rendering. This branch goes with https://github.com/ebi-gene-expression-group/atlas/pull/218. Here is only a tiny modification on the Json payload for null markerGenes.

### CODE REVIEW

**Comment by alfonsomunozpomer** [comment]
Also, the demo page `faceted-search-container-demo.html` is now breaking. Could you please fix it?

File: src/FilterList.js [hunk 1]
```diff
@@ -2,16 +2,84 @@ import React from 'react'
...
-const FilterList = ({filteredResults, resultsMessage, ResultsHeaderClass, ResultElementClass}) => {
-  const filteredElements = filteredResults.map((result) => result.element)
-
-  return (
-    <div>
-      <h4>{resultsMessage}</h4>
-      { filteredElements.length && ResultsHeaderClass && <ResultsHeaderClass /> }
-      { filteredElements.map((element, index) => <div key={index}><ResultElementClass {...element}/></div>) }
-    </div>
+function dynamicSort(property, sortState) {
```
Comment (alfonsomunozpomer) [inline src/FilterList.js]: We use arrow notation for functions, so instead of `function dynamicSort(property, sortState) {` we do `const dynamicSort = (property, sortState) => {`. Change this also in the returned funciton and in `titleCase` below.

File: src/FilterList.js [hunk 2]
```diff
@@ -2,16 +2,84 @@ import React from 'react'
...
+        { 
+          tableTitles.map((title, index) => {
+            const TitleDiv = tableTitleDivs[index]
+            return Object.keys(this.props.filteredResults[0].element).includes(titleCase(title)) ?
+              titleCase(title) === this.state.sortTitle ?
+                this.state.sortState ?
+                  <TitleDiv key={title} style={{opacity: 1}}> <p id={`selected`} onClick={this.sortTable}>{title} &darr;</p></TitleDiv> :
+                  <TitleDiv key={title} style={{opacity: 1}}> <p id={`selected`} onClick={this.sortTable}>{title} &uarr;</p></TitleDiv> 
+                :
+                <TitleDiv key={title}> <p id={`title`} onClick={this.sortTable}>{title} &uarr;</p></TitleDiv>
```
Comment (alfonsomunozpomer) [inline src/FilterList.js]: You are using an HTML entity for the upwards-pointing arrow and a Unicode escape sequence below for the downwards-pointing arrow. Use one or the other (I’d suggest you just copy and paste the character, since we can write code in UTF-8 now). I’d however use triangles, since those are most widely used in other tables: ▲, ▼.

File: src/FilterList.js [hunk 3]
```diff
@@ -2,16 +2,84 @@ import React from 'react'
...
+              :
+              <TitleDiv key={title}> <p id={`title`} onClick={this.sortTable}>{title}</p></TitleDiv> 
+          }) 
+        }
+      </ContainerDiv>
+
+    return TableHeader
+  }
+
+  sortTable(event){
```
Comment (alfonsomunozpomer) [inline src/FilterList.js]: Space before curly brace.

File: src/FilterList.js [hunk 4]
```diff
@@ -2,16 +2,84 @@ import React from 'react'
...
-
-  return (
-    <div>
-      <h4>{resultsMessage}</h4>
-      { filteredElements.length && ResultsHeaderClass && <ResultsHeaderClass /> }
-      { filteredElements.map((element, index) => <div key={index}><ResultElementClass {...element}/></div>) }
-    </div>
+function dynamicSort(property, sortState) {
+  const sortOrder = sortState ?  -1 : 1
+  return function (a,b) {
```
Comment (alfonsomunozpomer) [inline src/FilterList.js]: Space between arguments.

File: src/FilterList.js [hunk 5]
```diff
@@ -2,16 +2,84 @@ import React from 'react'
...
+    let result = (a[property] < b[property]) ? -1 : (a[property] > b[property]) ? 1 : 0
```
Comment (alfonsomunozpomer) [inline src/FilterList.js]: You can use `const` here.

File: src/FilterList.js [hunk 6]
```diff
@@ -2,16 +2,84 @@ import React from 'react'
...
+  }
+}
+
+function titleCase(str) {
+  const splitStr = str.toLowerCase().split(` `).map((string, index)=>
+    index === 0 ?
+      string.charAt(0).toLowerCase() + string.substring(1) :
+      string.charAt(0).toUpperCase() + string.substring(1)    
   )
+  return splitStr.join(``)
```
Comment (alfonsomunozpomer) [inline src/FilterList.js]: You can chain `join` after `map`.

File: src/FilterList.js [hunk 7]
```diff
@@ -2,16 +2,84 @@ import React from 'react'
...
+      sortState: escape(event.target.innerText.slice(-1)) === `%u2191`
+    })
+  }
+
+  render() {
+    const {filteredResults, resultsMessage, ResultsHeaderClass, ResultElementClass} = this.props
+
+    const filteredElements = filteredResults.map((result) => result.element)
+    const TableHeader = this.constructTableHeader(ResultsHeaderClass)
+    filteredElements.sort(dynamicSort(this.state.sortTitle, this.state.sortState))
```
Comment (alfonsomunozpomer) [inline src/FilterList.js]: Since we’re including lodash in the project, I suggest you use [`_.sortBy`](https://lodash.com/docs/4.17.11#sortBy) here with `reverse`. That involves writing less code and the performance hit is negligible.

File: src/FilterSidebar.js [hunk 8]
```diff
@@ -5,7 +5,7 @@ import _ from 'lodash'
 import CheckboxFacetGroup from './facetgroups/CheckboxFacetGroup'
 import MultiselectDropdownFacetGroup from './facetgroups/MultiselectDropdownFacetGroup'
 
-import {FacetPropTypes, FacetTooltipPropTypes} from './ResultPropTypes'
```
Comment (alfonsomunozpomer) [inline src/FilterSidebar.js]: Thanks for cleaning this up!

File: src/FilterList.js [hunk 9]
```diff
@@ -2,16 +2,84 @@ import React from 'react'
...
   )
+  return splitStr.join(``)
+}
+
+class FilterList extends React.Component {
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      sortTitle: `species`,
```
Comment (alfonsomunozpomer) [inline src/FilterList.js]: Initialize the `sortTitle` to e.g. the first field of the header, don’t assume any fields are present.

File: src/FilterList.js (hunk 7 above)
Comment (lingyun1010) [reply src/FilterList.js]: I do not think I can pass `this.state.sortTitle` into the sorting function of `_.sortBy`.

File: src/FilterList.js (hunk 2 above)
Comment (lingyun1010) [reply src/FilterList.js]: Excellent point!

**Comment by alfonsomunozpomer** [comment]
Oh no! Somewhere along the last set of commits sorting stopped working. Have a look at this two images:
**Sorted by number of assays**
image

**Sorted by title**
image

I have checked and commit 196246c56f3d6fbc7815a0041317a21e144e4a88 was working fine. Please review the code and add a test that checks that sorting works.

File: html/fetch-loader-demo.html [hunk 10]
```diff
@@ -5,6 +5,7 @@
 
     <link rel="stylesheet" href="https://ebi.emblstatic.net/web_guidelines/EBI-Framework/v1.3/css/ebi-global.css" type="text/css" media="all" />
     <link rel="stylesheet" href="https://ebi.emblstatic.net/web_guidelines/EBI-Icon-fonts/v1.3/fonts.css" type="text/css" media="all" />
+    <link rel="stylesheet" href="https://ebi.emblstatic.net/web_guidelines/EBI-Icon-fonts/v1.3/fonts.css" type="text/css" media="all">
```
Comment (alfonsomunozpomer) [inline html/fetch-loader-demo.html]: Why is this added twice?

File: src/FilterList.js [hunk 11]
```diff
@@ -1,17 +1,40 @@
...
-      <h4>{resultsMessage}</h4>
-      { filteredElements.length && ResultsHeaderClass && <ResultsHeaderClass /> }
-      { filteredElements.map((element, index) => <div key={index}><ResultElementClass {...element}/></div>) }
-    </div>
-  )
+    this.state = {
+      ascending: false
+    }
+
+    this.sortTable = this.sortTable.bind(this)
```
Comment (alfonsomunozpomer) [inline src/FilterList.js]: I think `sortTable` is not the best name since the results aren’t necessarily a table. Can you think of a better name?

File: html/fetch-loader-demo.html (hunk 10 above)
Comment (alfonsomunozpomer) [inline html/fetch-loader-demo.html]: This CSS has been added twice.
